### PR TITLE
ARROW-14193: [C++][Gandiva] Implement INSTR function

### PR DESCRIPTION
**Pull request by augustoasilva**
Returns the position of the first occurrence of _substr_ in _str_. Returns null if either of the arguments are null and returns 0 if substr could not be found in str.

### CODE REVIEW

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:805]: We should avoid using std::string as this does a memcpy. It would also do a malloc, malloc requires a lock impacting performance on multi-threaded machines.

You should write a loop like as follows:
for(int i = 0; i < (string_len - substring_len); i++) {
  if (string[i] == substring[0]) {
    // check if there is a complete match starting from this index
  }
}

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:800]: If the substring is a literal, you should be able to pre-process the substring to implement this faster. Can you open a separate ticket to handle a literal substring?

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:785]: Both of these tests match at the end. Can you add tests where the substring is at index 0 (the return value should be 1) and where the substring is somewhere in the middle?

Also, a test case where the string has most of the substring except the last character - for e.g. instr("hello world", "worldA") - this will catch off-by-1 errors in the code

File: cpp/src/gandiva/tests/projector_test.cc
Comment (vvellanki) [inline cpp/src/gandiva/tests/projector_test.cc:1635]: What is the expected output if the substr is not null, but the length is 0? Is this matching at index 1 or not matching?

The Hive documentation is not clear on this. Can you test how the Hive function behaves? My testing shows that an empty substring (not-null, but a substring of "") returns 1 for the INSTR(column, '')

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:804]: My testing shows that this should return 1

File: cpp/src/gandiva/function_registry_string.cc
Comment (vvellanki) [inline cpp/src/gandiva/function_registry_string.cc:412]: Why does this function need context? I also dont see this function return errors

Do we need the kNeedsContext and kCanReturnErrors flags?

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs.cc:805]: Ok, fixed it

File: cpp/src/gandiva/tests/projector_test.cc
Comment (augustoasilva) [reply cpp/src/gandiva/tests/projector_test.cc:1635]: Yest, I tested it here too, and it now returns 1

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs_test.cc:804]: Fixed it

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs_test.cc:785]: I added this test case, and it passed!

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs.cc:801]: Yes, no problem. Should I open it before or after this one is merged?

File: cpp/src/gandiva/function_registry_string.cc
Comment (augustoasilva) [reply cpp/src/gandiva/function_registry_string.cc:412]: This one I was waiting for review, so I just left kNeedsContext, if I remove it the projector test will fail

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (vvellanki) [reply cpp/src/gandiva/gdv_function_stubs.cc:801]: After this one is merged is fine

File: cpp/src/gandiva/function_registry_string.cc
Comment (vvellanki) [reply cpp/src/gandiva/function_registry_string.cc]: This is not going to allocate memory. So, I dont understand why it needs context

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:802]: This code is going to run for all the rows in the table. So, it is important to avoid if-statements in the code.

How about the following:
if (substring_len == 0) { return 1; } // I see instr('', '') returning 1. So, there is no need to check for string_len

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:817]: This is incorrect - if string = "Hello World" and substring = "HW", this implementation will return 1

You should change this to:
for(int i = 0; i < string_len; i++) {
  if (string[i] == substring[0] && stringMatches(string + 1, substring + 1, substring_len - 1)) {
    return (i + 1);
  }
}

return 0;

stringMatches(const char *string, const char *substring, int substring_len) {
  for(int i = 0; i < substring_len; i++) {
    if (string[i] != substring[i]) {
      return false;
    }
  }

  return true;
}

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs.cc:802]: Done

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs.cc]: Done

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:808]: Please use memcmp instead - https://www.cplusplus.com/reference/cstring/memcmp/

You are essentially comparing the next substring_len - 1 bytes of both the pointers

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:807]: The condition (i < string_len) is incorrect. Note that you are comparing the next N bytes when the first byte matches. If there is a match at the last byte, you will access beyond the end here and cause issues

This should be changed to (i < string_len - substring_len)

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:787]: s2="apple" should return 1

File: cpp/src/gandiva/string_utils.cc
Comment (vvellanki) [inline cpp/src/gandiva/string_utils.cc:22]: This is not required, please use memcmp() instead

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs.cc]: Fixed

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs.cc]: It is using memcmp now

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs_test.cc:787]: It returns 1 now

File: cpp/src/gandiva/function_registry_string.cc
Comment (augustoasilva) [reply cpp/src/gandiva/function_registry_string.cc]: Removed kNeedsContext flag

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:806]: Please handle the case where string_len is less than substring_len

Comment (vvellanki) [reply cpp/src/gandiva/gdv_function_stubs.cc:806]: Also, add test cases for string_len < substring_len

Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs.cc:810]: Done!

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs.cc:806]: What happens if string_len == substring_len? Looks like we return 0 always - the for loop goes from 0 to 0

Can you fix this case?

Also, can this function be moved to precompiled and use FORCE_INLINE?

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (vvellanki) [inline cpp/src/gandiva/gdv_function_stubs_test.cc:778]: Please add a test where string = "Hello" and substring = "Hello". Another where string = "Hello", substring = "Hell"

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs.cc]: Fixed this if clause now, it will handle case when string_len == substring_len and when string_len < substring_len

File: cpp/src/gandiva/gdv_function_stubs_test.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs_test.cc]: Added

File: cpp/src/gandiva/precompiled/string_ops.cc
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/string_ops.cc:2209]: Let the input be: "Hello" with string_len = 5
substring = "lower" with substring_len = 5

end_idx = 5 in this case

For i = 2, string[i] = 'l' = substring[0]. The memcmp() is going to access string + i + substring_len, which is string + 2 + 5 = string + 7. string + 7 is out of range of the input. We cannot do this, it will cause issues

Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops.cc:2209]: Change the memcmp to strncmp to prevent this as strncmp as it has the advantage of its stop-on-NUL behavior

File: cpp/src/gandiva/precompiled/string_ops.cc
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/string_ops.cc:2476]: if (string_len < substring_len) - shouldn't this return 0 without any further checking?

File: cpp/src/gandiva/precompiled/string_ops.cc
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/string_ops.cc:2480]: You cannot use strncmp() since this terminates if it sees a '\0' char. You should use memcmp() with the appropriate size

File: cpp/src/gandiva/precompiled/string_ops.cc
Comment (vvellanki) [reply cpp/src/gandiva/precompiled/string_ops.cc:2476]: else, your end_idx = string_len - substring_len;

And, your for loop should go from i = 0; to i <= end_idx

File: cpp/src/gandiva/precompiled/string_ops.cc
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops.cc:2632]: The substring_len is ok now, the comparing was worng, memcmp is now ok

File: cpp/src/gandiva/precompiled/string_ops.cc
Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops.cc:2628]: Fixed

File: cpp/src/gandiva/precompiled/string_ops_test.cc
Comment (vvellanki) [inline cpp/src/gandiva/precompiled/string_ops_test.cc:2178]: Did this test pass with this version of the code? If so, the code has another bug around the length

Comment (vvellanki) [reply cpp/src/gandiva/precompiled/string_ops_test.cc:2178]: We should have more negative tests with the following:
- The substring's first 2 chars only match. The 3rd char doesn't match
- Test cases with this matching substring at the beginning of string, middle of the string and end of string

Comment (augustoasilva) [reply cpp/src/gandiva/precompiled/string_ops_test.cc:2178]: Yest the first test passed, and I added more test cases

**Review by vvellanki — COMMENTED** [review]
Looks good. Please address the comment - and I'll approve the change

File: cpp/src/gandiva/gdv_function_stubs.cc
Comment (augustoasilva) [reply cpp/src/gandiva/gdv_function_stubs.cc]: Is not necessary, so removed

**Review by vvellanki — APPROVED** [review]

**Pull request closed without merging**
